<a href="https://colab.research.google.com/github/agmalaga2020/PolicySpace2_Spanish_data/blob/main/cifras_poblacion_municipio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# INICIO

- 1º buscamos fuente a explorar
https://uoc0-my.sharepoint.com/:x:/g/personal/agmalaga_uoc_edu/EVrtEuZkGqxMsXz3ENcmP8sB8w9sxyyFWBX3pDP282BDaQ?e=hZX8jz

_______________________________________________________________

- 2º `CARGAMOS` DATAFRAME ORIGINAL del proyecto **github**
https://github.com/agmalaga2020/PolicySpace2-agmalaga/tree/master/input

_______________________________________________________________
- 3º Realizamos analisis exploratorio en google colab.
_______________________________________________________________

- 4º descargamos datos y creamos informe.
_______________________________________________________________

- 5º actualizar LINK de la columna `NOTEBOOK`
https://uoc0-my.sharepoint.com/:x:/g/personal/agmalaga_uoc_edu/EVrtEuZkGqxMsXz3ENcmP8sB8w9sxyyFWBX3pDP282BDaQ?e=hZX8jz

_______________________________________________________________

ESQUEMA DE PROCESO

https://whiteboard.office.com/me/whiteboards/p/c3BvOmh0dHBzOi8vdW9jMC1teS5zaGFyZXBvaW50LmNvbS9wZXJzb25hbC9hZ21hbGFnYV91b2NfZWR1/b!r4X97qYbxUyQGQNSpn-bfrNi376yDThFsrR2Kul1sxTx_nU34TfBSpdz-6pNWzvq/01E765GIPQJMS23ADWVBEID5XGUDPDSQNZ?source=applauncher&auth_upn=agmalaga%40uoc.edu

### pop_men_2000.csv "ejemplo"


https://github.com/agmalaga2020/PolicySpace2-agmalaga/blob/master/input/pop_men_2000.csv

In [17]:
import requests
import pandas as pd
import io

# Update the URL to point to the raw CSV data on GitHub
url = 'https://raw.githubusercontent.com/agmalaga2020/PolicySpace2-agmalaga/master/input/pop_men_2000.csv'
s = requests.get(url).content

# Read the CSV data using ';' as the separator
df = pd.read_csv(io.StringIO(s.decode('utf-8')), sep=';')

print(df.head())
print("el dataframe original de brasil contiene un total de municipios:", df.cod_mun.unique().shape[0])

   cod_mun     0     1     2     3     4     5     6     7     8  ...  91  92  \
0  2700102   228   205   252   255   259   255   233   219   226  ...   1   1   
1  2700201   210   223   202   229   194   222   221   227   212  ...   0   0   
2  2700300  2140  2054  2052  2098  2210  2241  2123  2023  2071  ...  12  16   
3  2700409   508   508   544   520   586   587   535   445   511  ...   4   2   
4  2700508   175   161   164   159   177   179   142   132   149  ...   1   1   

   93  94  95  96  97  98  99  100  
0   3   1   1   1   0   1   0    5  
1   1   3   1   0   0   0   0    0  
2   7   9   7   8   9   4   2   13  
3   0   3   2   1   2   1   0    6  
4   2   1   0   0   0   0   0    1  

[5 rows x 102 columns]
el dataframe original de brasil contiene un total de municipios: 5485


# ESPAÑA

tenemos que extraer tabla a tabla por municipio para finalmente transformar los datos para conseeguir el mismo resultado.

### Población por sexo, municipios y edad (año a año) - Nacional

https://www.ine.es/jaxiT3/Tabla.htm?t=33575&L=0

# ocupa mas de 1 gb

### extraemos las tablas

In [ ]:
import requests
import pandas as pd

# Código de la tabla
table_code = "33575"
url_csv = f"https://servicios.ine.es/wstempus/csv/ES/DATOS_TABLA/{table_code}?nult=999"

response = requests.get(url_csv)

if response.status_code == 200:
    csv_data = response.content.decode('utf-8')

    # Guardar el archivo CSV con un nombre adecuado
    file_name = "tabla_33847.csv"
    with open(file_name, "w", encoding="utf-8") as f:
        f.write(csv_data)
    print(f"Archivo CSV guardado como '{file_name}'")
else:
    print("Error en la petición: ❌", response.status_code)

# Leer el CSV
df = pd.read_csv(file_name, sep='\t', encoding='utf-8')

# Mostrar las primeras filas
print(df.head())
print("Tabla actualizada conseguida ✅")

try:
  df.to_csv('/content/33575.csv', index=False)
  print("✅ Copia de df_nuevo_filtrado_2 guardada en '/content/33575.csv'")
except NameError:
  print("⚠️ df no está definido. No se puede guardar la copia.")
print("Se ha guardado una copia del DataFrame df.")


In [ ]:
import pandas as pd

# Cargamos el DataFrame desde el archivo CSV
df = pd.read_csv('/content/33575.csv')

# Mostramos las primeras filas del DataFrame
print(df.head())


In [46]:
def print_unique_values_and_len(df):
  """Imprime los valores únicos de cada columna en un DataFrame y su longitud.
  Args:
    df: El DataFrame de pandas.
  """
  for column in df.columns:
    unique_values = df[column].unique()
    nan_count = df[column].isna().sum()
    print(f"Valores únicos de la columna '{column}':")
    print(unique_values)
    print(f"Longitud de valores únicos: {len(unique_values)}")
    print(f"Número de NaN: {nan_count}")
    print(f"Tipo de dato de la columna: {df[column].dtype}")
    print()

print_unique_values_and_len(df)

Valores únicos de la columna 'sexo':
['Hombres' 'Mujeres']
Longitud de valores únicos: 2
Número de NaN: 0
Tipo de dato de la columna: object

Valores únicos de la columna 'edad':
[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71
 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95
 96 97 98 99]
Longitud de valores únicos: 100
Número de NaN: 0
Tipo de dato de la columna: int64

Valores únicos de la columna 'periodo':
['1 de enero de 2022' '1 de enero de 2021' '1 de enero de 2020'
 '1 de enero de 2019' '1 de enero de 2018' '1 de enero de 2017'
 '1 de enero de 2016' '1 de enero de 2015' '1 de enero de 2014'
 '1 de enero de 2013' '1 de enero de 2012' '1 de enero de 2011'
 '1 de enero de 2010' '1 de enero de 2009' '1 de enero de 2008'
 '1 de enero de 2007' '1 de enero de 2006' '1 de enero de 2005'
 '1 de 

✅ Resultado del análisis de valores únicos:

- Todo está limpio: no hay `NaN`, y las categorías han quedado bien separadas.
- `sexo`: 2 valores (`Hombres`, `Mujeres`).
- `edad`: 101 edades entre `0 años` y `100 y más años`.
- `periodo`: 20 años (2003 a 2022).
- `Total`: 7.657 valores únicos, pero aún en formato texto con posibles comas o puntos.
- `provincia_code` y `provincia_name`: solo un valor (`28`, `Madrid`).
- `municipio_code` y `municipio_name`: 179 municipios distintos.

¿Convertimos ahora `Total` a formato numérico (float) para análisis y agregaciones?